arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/stream/Scope.kt

```diff
@@ -174,19 +166,16 @@ class Scope private constructor(
     fr: suspend () -> R,
     release: suspend (R, ExitCase) -> Unit
   ): Either<Throwable, R> {
-    val resource = ScopedResource()
-    return when (val eith = Either.catch(fr)) {
-      is Either.Right ->
-        when (resource.acquired { ex -> release(eith.b, ex) }) {
-          is Either.Right -> {
-            register(resource)
-            eith.b.right()
-          }
-          is Either.Left ->
-            Either.Left(eith.swap().getOrElse { AcquireAfterScopeClosed })
-        }
-
-      is Either.Left -> eith
+    val conn = coroutineContext.connection()
+    val scope = ScopedResource()
+    return Either.catch(fr).flatMap { resource ->
+      scope.acquired { ex: ExitCase -> release(resource, ex) }.map { registered ->
+        state.modify {
+          if (conn.isCancelled() && registered) Pair(it, suspend { release(resource, ExitCase.Cancelled) })
```

Comment:
We need to check for cancellation while registering the resource into the scope, if while registering the resource a cancellation signal occurs than we need to release the resource immediately with `ExitCase.Cancelled`.

Reply:
Same here, conn could be inlined